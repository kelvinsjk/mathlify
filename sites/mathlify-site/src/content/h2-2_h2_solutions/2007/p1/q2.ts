import { Answer } from '$lib/classes/answer';
import { QED } from '$lib/utils/typesetting/utils';
import { Polynomial } from 'mathlify';
import { generateDomain } from '$content/h2-1_h2_learn/01_functions/02_inverse/02_domain/02_practice';
import { improperInverse } from '$content/h2-1_h2_learn/01_functions/02_inverse/03_formula/02_practice';
import { compositeExists } from '$content/h2-1_h2_learn/01_functions/03_composite/01_existence/02_practice';
import { compositeFormula } from '$content/h2-1_h2_learn/01_functions/03_composite/02_formula/02_practice';

export const answer = new Answer();

// ! Complete
// f(x) = 1/(x-3)
// g(x) = x^2

const [a, b, c] = [-3, 0, 1];
const f = new Polynomial([1, a]).reciprocal();
const g = Polynomial.ofDegree(2);
const fState: Parameters<typeof improperInverse>[0] = {
	fnType: 'frac',
	a,
	b,
	c,
	restriction: false,
	unknownConstants: false,
	definition: false
};
const gState = {
	fnType: 'quadratic',
	a: 0,
	b: 0,
	c: 0,
	restriction: false,
	unknownConstants: false
} as const;
const compositeStateFg: Parameters<typeof compositeExists>[0] = {
	f: fState,
	g: gState,
	fg: true
};

// a
{
	const { ans: ans2, soln: soln2 } = compositeExists(compositeStateFg, [f, g]);
	const { ans, soln } = compositeFormula(
		[f, generateDomain(fState), g, generateDomain(gState)],
		false,
		true,
		{ ansInline: true, QED: true }
	);
	answer.addPart(ans + '\\\n' + ans2, soln + '\n\n' + soln2);
}
// b
{
	const { ans, soln } = improperInverse(fState, {
		definition: false,
		rhs: f,
		QED,
		omitReasoningInAns: true
	});
	answer.addPart(ans, soln);
}
