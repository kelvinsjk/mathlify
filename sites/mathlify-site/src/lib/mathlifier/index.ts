/**
 * generates djot string,
 * using JavaScript's template literal syntax to handle mathematical expressions (both static and dynamic)
 *
 * math: starts with ${}, terminates with \n,
 * display: starts with $${}, terminates with \n\n,
 * display env: starts with $${'align'}, etc.
 * text: starts with @${}, terminates immediately
 */
export function mathlify(strings: TemplateStringsArray, ...values: unknown[]): string {
	return mathlifyGen(modules)(strings, ...values);
}

const modules: Modules = {
	math: (x: string) => `$\`${x}\``,
	display,
	mathEnvs: {
		equation: (x: string) => display(insertEnv('equation', x)),
		align: (x: string) => display(insertEnv('align', x)),
		gather: (x: string) => display(insertEnv('gather', x)),
		alignat: (x: string) => display(insertEnv('alignat', x)),
		equationStar: (x: string) => display(insertEnv('equation*', x)),
		alignStar: (x: string) => display(insertEnv('align*', x)),
		gatherStar: (x: string) => display(insertEnv('gather*', x)),
		alignatStar: (x: string) => display(insertEnv('alignat*', x)),
	},
};

function insertEnv(env: string, content: string, args = ''): string {
	if (env === 'alignat' || env === 'alignat*') {
		let firstLine = content.split('\n')[0];
		if (!firstLine) {
			firstLine = content.split('\n')[1];
		}
		const twoN = firstLine.split('&').length;
		if (twoN % 2 !== 0)
			console.warn(`alignat should have an even number of columns, but got ${twoN} instead`);
		args = `{${Math.floor(twoN / 2)}}`;
	}
	// todo: handle this with regex
	while (content.endsWith('\n') || content.endsWith('\t') || content.endsWith(' ')) {
		content = content.slice(0, content.length - 1);
	}
	return `\\begin{${env}}${args}${content}\n\\end{${env}}`;
}

function display(x: string): string {
	return `$$\`${x}\``;
}

enum Modes {
	text, // 0
	math, // 1
	display, // 2
	mathEnv, // 3
}

export interface Modules {
	math: (x: string) => string;
	display: (x: string) => string;
	mathEnvs?: {
		equation?: (x: string) => string;
		align?: (x: string) => string;
		gather?: (x: string) => string;
		alignat?: (x: string) => string;
		equationStar?: (x: string) => string;
		alignStar?: (x: string) => string;
		gatherStar?: (x: string) => string;
		alignatStar?: (x: string) => string;
	};
}

enum MathEnvs {
	equation,
	equationStar,
	align,
	alignStar,
	gather,
	gatherStar,
	alignat,
	alignatStar,
}

const mathEnvsKeys = [
	'equation',
	'equation*',
	'align',
	'align*',
	'gather',
	'gather*',
	'alignat',
	'alignat*',
] as const;
function isMathEnv(x: string): x is (typeof mathEnvsKeys)[number] {
	return (mathEnvsKeys as readonly string[]).includes(x);
}

const envToMode: Record<(typeof mathEnvsKeys)[number], MathEnvs> = {
	equation: MathEnvs.equation,
	'equation*': MathEnvs.equationStar,
	align: MathEnvs.align,
	'align*': MathEnvs.alignStar,
	gather: MathEnvs.gather,
	'gather*': MathEnvs.gatherStar,
	alignat: MathEnvs.alignat, // triggered by $${'alignat'}{x}
	'alignat*': MathEnvs.alignatStar, // triggered by $${'alignat'}{x}
};

interface MathEnvOptions {
	mathEnv: MathEnvs;
	eqnColumns?: number;
}

// TODO: alignat cols

export function mathlifyGen(
	modules: Modules,
): (strings: TemplateStringsArray, ...values: unknown[]) => string {
	return (strings: TemplateStringsArray, ...values: unknown[]) => {
		let currentModeString = '';
		let output = '';
		let mode = Modes.text;
		let mathEnvOptions: MathEnvOptions = { mathEnv: MathEnvs.equation };
		strings.forEach((str, i) => {
			const nextVal = values[i] ?? '';
			if (mode === Modes.text) {
				if (str.endsWith('@')) {
					// continues text mode
					currentModeString += `${str.slice(0, str.length - 1)}${nextVal}`;
				} else {
					// starts new text environment
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						str,
						'',
						nextVal,
						currentModeString,
					);
				}
			} else if (mode === Modes.math) {
				// checks for \n or \r\n non-greedily
				const regex = /([^]*?)(\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end math mode
					const [, before, newline, after] = match;
					output += `${before}`;
					if (output) {
						currentModeString += modules.math(output);
					}
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						currentModeString,
					);
				} else {
					// continue math mode
					if (str.endsWith(`@`)) {
						// TODO: check this
						output += `${str.slice(0, str.length - 1)}`;
					} else {
						output += `${str}${nextVal}`;
					}
					if (i === strings.length - 1) {
						// final string
						currentModeString += modules.math(output);
					}
				}
			} else if (mode === Modes.display) {
				// checks for \n\n or \r\n\r\n within str. group into (#1)(#2) where #2 = \n or \r\n and it is done non-greedily
				const regex = /([^]*?)(\r?\n[ \t]*\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end display mode
					const [, before, newline, after] = match;
					output += before;
					if (output) {
						currentModeString += modules.display(output);
					}
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						currentModeString,
					);
				} else {
					// continue display mode
					if (str.endsWith(`@`)) {
						// TODO: check this
						output += `${str.slice(0, str.length - 1)}`;
					} else {
						output += `${str}${nextVal}`;
					}
					if (i === strings.length - 1) {
						// final string
						currentModeString += modules.display(output);
					}
				}
			} else if (mode === Modes.mathEnv) {
				// checks for \n\n or \r\n\r\n within str. group into (#1)(#2) where #2 = \n or \r\n and it is done non-greedily
				const regex = /([^]*?)(\r?\n[ \t]*\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end mathEnv mode
					const [, before, newline, after] = match;
					output += before;
					if (output) {
						currentModeString += insertMathEnv(mathEnvOptions, output, modules);
					}
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						currentModeString,
					);
				} else {
					// continue mathEnv mode
					if (str.endsWith(`@`)) {
						// TODO: check this.
						output += `${str.slice(0, str.length - 1)}`;
					} else {
						output += `${str}${nextVal}`;
					}
					if (i === strings.length - 1) {
						// final string
						currentModeString += insertMathEnv(mathEnvOptions, output, modules);
					}
				}
			}
		});
		return currentModeString;
	};
}

// returns [Mode, acc, curr, {mathEnv, eqnColumns?}]
function startNewEnv(
	after: string,
	newline: string,
	nextVal: unknown,
	curr: string,
): [Modes, string, string, MathEnvOptions] {
	const defaultMathEnvOptions = { mathEnv: MathEnvs.equation };
	let mathEnvOptions: MathEnvOptions = defaultMathEnvOptions;
	if (after.endsWith('$')) {
		let x = `${nextVal}`;
		let mode = Modes.display;
		if (isMathEnv(x)) {
			// new math env
			mode = Modes.mathEnv;
			mathEnvOptions = { mathEnv: envToMode[x] };
			//TODO: handle alignat cols
			x = '';
		}
		// new display mode
		return [mode, x, curr + `${newline}${after.slice(0, after.length - 1)}`, mathEnvOptions];
	} else if (after.endsWith('@')) {
		// new text mode
		return [
			Modes.text,
			'',
			curr + `${newline}${after.slice(0, after.length - 1)}${nextVal}`,
			defaultMathEnvOptions,
		];
	} else {
		// new math mode
		return [Modes.math, `${nextVal}`, curr + newline + after, defaultMathEnvOptions];
	}
}

function insertMathEnv(mathEnvOptions: MathEnvOptions, acc: string, modules: Modules): string {
	const mathEnv = mathEnvOptions.mathEnv;
	if (mathEnv === MathEnvs.equation) {
		return modules.mathEnvs?.equation ? modules.mathEnvs.equation(acc) : acc;
	} else if (mathEnv === MathEnvs.equationStar) {
		return modules.mathEnvs?.equationStar ? modules.mathEnvs.equationStar(acc) : acc;
	} else if (mathEnv === MathEnvs.align) {
		return modules.mathEnvs?.align ? modules.mathEnvs.align(acc) : acc;
	} else if (mathEnv === MathEnvs.alignStar) {
		return modules.mathEnvs?.alignStar ? modules.mathEnvs.alignStar(acc) : acc;
	} else if (mathEnv === MathEnvs.gather) {
		return modules.mathEnvs?.gather ? modules.mathEnvs.gather(acc) : acc;
	} else if (mathEnv === MathEnvs.gatherStar) {
		return modules.mathEnvs?.gatherStar ? modules.mathEnvs.gatherStar(acc) : acc;
	} else if (mathEnv === MathEnvs.alignat) {
		return modules.mathEnvs?.alignat ? modules.mathEnvs.alignat(acc) : acc;
	} else if (mathEnv === MathEnvs.alignatStar) {
		return modules.mathEnvs?.alignatStar ? modules.mathEnvs.alignatStar(acc) : acc;
	}
	console.warn(`mathEnv ${mathEnv} not found`);
	return acc;
}
