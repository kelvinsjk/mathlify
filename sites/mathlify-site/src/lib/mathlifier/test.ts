import { mathlify } from './';

const str = mathlify`This is inline math: ${'x+2'}xyz${'k'}
and this is text.

This is display math: $${'x+2'}xyz
abcd
${'k'}

This is aligned:
$${'align'}
  x+2 &= ${'k'} \\\\
  y &= ${'z'}`;

console.log(str);
