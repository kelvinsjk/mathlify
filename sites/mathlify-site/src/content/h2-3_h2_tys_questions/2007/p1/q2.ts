/* The question here is the copyright of 
- Singapore Examinations and Assessment Board
- Cambridge Assessment International Education
- Ministry of Education, Singapore
- Government of Singapore

We provide this question on Mathlify under fair use for archival purposes.
These are questions more than 10 years old and are difficult to obtain
commercially.

For the newer questions, please purchase a copy of the
"ten year series" from authorized publishers and distributors.
*/

import { Polynomial } from 'mathlify';
import { Question } from '$lib/classes/question';
import { mathlifierDj as mathlifier } from 'mathlifier';

export const question = new Question();

//f(x) = 1/(x-3)
//g(x) = x^2

const a = -3;
const f = new Polynomial([1, a]).reciprocal();
const g = Polynomial.ofDegree(2);

question.addBody(
	mathlifier`Functions ${'f'}
and ${'g'}
are defined by
$${'align*'}
&f: x \\mapsto ${f}  \\quad && \\text{for } x \\in \\mathbb{R}, x \\neq 3, \\\\
&g: x \\mapsto ${g}  \\quad && \\text{for } x \\in \\mathbb{R}.
`
);
// a
{
	question.addPart(
		mathlifier`Only one of the
composite functions ${'fg'}
and ${'gf'}
exists. Give a definition (including the domain)
of the composite that exists, and explain why
the other composite does not exist.`,
		{ marks: 3 }
	);
}
// b
{
	question.addPart(
		mathlifier`Find ${'f^{-1}(x)'}
and state the domain of ${'f^{-1}'}.`,
		{ marks: 3 }
	);
}
