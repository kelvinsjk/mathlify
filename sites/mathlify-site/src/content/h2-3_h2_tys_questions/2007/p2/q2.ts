/* The question here is the copyright of 
- Singapore Examinations and Assessment Board
- Cambridge Assessment International Education
- Ministry of Education, Singapore
- Government of Singapore

We provide this question on Mathlify under fair use for archival purposes.
These are questions more than 10 years old and are difficult to obtain
commercially.

For the newer questions, please purchase a copy of the
"ten year series" from authorized publishers and distributors.
*/

import { Question } from '$lib/classes/question';
import { mathlifierDj as mathlifier } from 'mathlifier';

export const question = new Question();

question.addBody(mathlifier`Out of syllabus (method of differences).`, { marks: 11 });
