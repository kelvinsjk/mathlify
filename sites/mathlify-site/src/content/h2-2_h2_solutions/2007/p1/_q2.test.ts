import { expect, it } from 'vitest';
import { answer } from './q2';
import { mathlifierDj as mathlifier } from 'mathlifier';

const ans = answer.answer;
const actual: typeof ans = {
	parts: [
		{
			body:
				mathlifier`${{}}gf:x\\mapsto \\frac{1}{\\left( x - 3 \\right)^2}, \\quad \\allowbreak  {x \\in \\mathbb{R}, x \\neq 3}.` +
				'\\\n' +
				mathlifier`${'fg'}
does not exist as ${{}} {R_{g} \\not  \\subseteq D_{f}.}`
		},
		{
			body: mathlifier`${{}} f^{-1}(x) = \\frac{1 + 3x}{x}.
\\
${{}} D_{f^{-1}}  = \\left( -\\infty, 0 \\right) \\cup \\left( 0, \\infty \\right).`
		}
	]
};
it('2007p1q2', () => {
	expect(ans).toEqual(actual);
	expect(answer).toMatchSnapshot();
});
