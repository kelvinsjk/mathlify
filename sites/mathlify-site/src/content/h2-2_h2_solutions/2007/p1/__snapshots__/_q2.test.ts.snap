// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`2007p1q2 1`] = `
Answer {
  "answer": {
    "parts": [
      {
        "body": "$\` gf:x\\mapsto \\frac{1}{\\left( x - 3 \\right)^2}, \\quad \\allowbreak  {x \\in \\mathbb{R}, x \\neq 3}. \`\\
$\` fg \`
does not exist as $\`  {R_{g} \\not  \\subseteq D_{f}.} \`",
      },
      {
        "body": "$\`  f^{-1}(x) = \\frac{1 + 3x}{x}. \`
\\
$\`  D_{f^{-1}}  = \\left( -\\infty, 0 \\right) \\cup \\left( 0, \\infty \\right). \`",
      },
    ],
  },
  "solution": {
    "parts": [
      {
        "body": "$$\` \\begin{align*}
gf(x) &= g \\left( \\frac{1}{x - 3} \\right)
\\\\ &= \\left( \\frac{1}{x - 3} \\right)^2
	\\\\ &= \\frac{1}{\\left( x - 3 \\right)^2}
\\end{align*} \`
$$\`  gf: x \\mapsto \\frac{1}{\\left( x - 3 \\right)^2}, \\quad x \\in \\mathbb{R}, x \\neq 3 \\; \\QED  \`

$$\` \\begin{align*}
R_{g} &= \\left[ 0, \\infty \\right) \\\\
D_{f} &= \\left( -\\infty, 3 \\right) \\cup \\left( 3, \\infty \\right)
\\end{align*} \`

Since $\`  R_{g} \\not  \\subseteq D_{f}, \`
the composite function $\` fg \`
does not exist $\` \\QED  \`
",
      },
      {
        "body": "$$\` \\begin{gather*} \\text{Let }  y = \\frac{1}{x - 3}
	\\\\ y\\left( x - 3 \\right) = 1
	\\\\ yx - 3y = 1
	\\\\ yx = 1 + 3y
	\\\\ x = \\frac{1 + 3y}{y}
\\end{gather*} \`

$$\`  f^{-1}(x) = \\frac{1 + 3x}{x} \\QED  \`

$$\` \\begin{align*} D_{f^{-1}} &= R_{f} \\\\ &= \\left( -\\infty, 0 \\right) \\cup \\left( 0, \\infty \\right) \\QED
\\end{align*} \`",
      },
    ],
  },
}
`;
